### PR TITLE
Optimize memory allocation of Lexer based on the stack

### PR DESCRIPTION
**Pull request by clover2123**
* remove default memory pool for ScannerResult
* each ScannerResult is now allocated on the stack if it is really necessary
* reduce the reference count overhead for each ScannerResult
* ALLOC_TOKEN macro is newly added for allocation of each ScannerResult token

Performance Result)
Octane : 4~5% improvement on average

Signed-off-by: HyukWoo Park <hyukwoo.park@samsung.com>

### CODE REVIEW

File: src/parser/Lexer.cpp [hunk 1]
```diff
@@ -857,9 +824,9 @@ uint16_t Scanner::octalToDecimal(char16_t ch, bool octal)
     return octal ? code : NON_OCTAL_VALUE;
 };
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanPunctuator(char16_t ch)
+void Scanner::scanPunctuator(Scanner::ScannerResult* token, char16_t ch)
 {
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 2]
```diff
@@ -1083,10 +1050,9 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanPunctuator(char16_t ch)
     }
 
     token->valuePunctuatorKind = kind;
-    return token;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanHexLiteral(size_t start)
+void Scanner::scanHexLiteral(Scanner::ScannerResult* token, size_t start)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 3]
```diff
@@ -1123,14 +1089,14 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanHexLiteral(size_t start)
...
+        token->setResult(this, Token::NumericLiteralToken, numberDouble, this->lineNumber, this->lineStart, start, this->index);
     } else {
         ASSERT(numberDouble == 0.0);
-        return adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index));
+        token->setResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index);
     }
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanBinaryLiteral(size_t start)
+void Scanner::scanBinaryLiteral(Scanner::ScannerResult* token, size_t start)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 4]
```diff
@@ -1158,10 +1124,10 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanBinaryLiteral(size_t start)
         }
     }
 
-    return adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index));
+    token->setResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index);
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanOctalLiteral(char16_t prefix, size_t start)
+void Scanner::scanOctalLiteral(Scanner::ScannerResult* token, char16_t prefix, size_t start)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 5]
```diff
@@ -1209,7 +1173,7 @@ bool Scanner::isImplicitOctalLiteral()
     return true;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanNumericLiteral()
+void Scanner::scanNumericLiteral(Scanner::ScannerResult* token)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 6]
```diff
@@ -1403,7 +1364,7 @@ bool Scanner::isFutureReservedWord(const StringView& id)
     return false;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanTemplate(bool head)
+void Scanner::scanTemplate(Scanner::ScannerResult* token, bool head)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 7]
```diff
@@ -1616,7 +1577,7 @@ String* Scanner::scanRegExpFlags()
     return new UTF16String(flags.data(), flags.length());
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanRegExp()
+void Scanner::scanRegExp(Scanner::ScannerResult* token)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 8]
```diff
@@ -1802,25 +1762,25 @@ static ALWAYS_INLINE KeywordKind isKeyword(const StringBufferAccessData& data)
     return NotKeyword;
 }
 
-ALWAYS_INLINE PassRefPtr<Scanner::ScannerResult> Scanner::scanIdentifier(char16_t ch0)
+ALWAYS_INLINE void Scanner::scanIdentifier(Scanner::ScannerResult* token, char16_t ch0)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.h [hunk 9]
```diff
@@ -253,75 +272,84 @@ class Scanner : public gc {
...
-            , startWithZero(false)
-            , octal(false)
-            , plain(plain)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueStringLiteralData(valueString)
+        void setResult(Scanner* scanner, Token type, size_t lineNumber, size_t lineStart, size_t start, size_t end)
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT scanner

File: src/parser/Lexer.h [hunk 10]
```diff
@@ -253,75 +272,84 @@ class Scanner : public gc {
...
-            , startWithZero(false)
-            , octal(false)
-            , plain(false)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueNumber(value)
+        void setResult(Scanner* scanner, Token type, StringView* valueString, size_t lineNumber, size_t lineStart, size_t start, size_t end, bool plain)
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT scanner

File: src/parser/Lexer.h [hunk 11]
```diff
@@ -253,75 +272,84 @@ class Scanner : public gc {
...
-            , startWithZero(false)
-            , octal(false)
-            , plain(false)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueTemplate(value)
+        void setResult(Scanner* scanner, Token type, double value, size_t lineNumber, size_t lineStart, size_t start, size_t end)
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT scanner

File: src/parser/Lexer.h [hunk 12]
```diff
@@ -253,75 +272,84 @@ class Scanner : public gc {
...
+            this->plain = false;
+            this->hasKeywordButUseString = true;
+            this->lineNumber = lineNumber;
+            this->lineStart = lineStart;
+            this->start = start;
+            this->end = end;
+            this->valueNumber = value;
+        }
+
+        void setResult(Scanner* scanner, Token type, ScanTemplteResult* value, size_t lineNumber, size_t lineStart, size_t start, size_t end)
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT scanner

Comment (yichoi) [reply src/parser/Lexer.h]: ASSERT value

File: src/parser/esprima_cpp/esprima.cpp [hunk 13]
```diff
@@ -388,13 +386,13 @@ class Parser : public gc {
 
         // this->sourceType = (options && options.sourceType == 'module') ? 'module' : 'script';
         this->sourceType = Script;
-        this->lookahead = nullptr;
+        //this->lookahead = nullptr;
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 14]
```diff
@@ -517,7 +515,7 @@ class Parser : public gc {
     }
 
     // Throw an exception because of an unexpected token.
-    void throwUnexpectedToken(RefPtr<Scanner::ScannerResult> token, const char* message = nullptr)
+    void throwUnexpectedToken(Scanner::ScannerResult* token, const char* message = nullptr)
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT token

File: src/parser/esprima_cpp/esprima.cpp [hunk 15]
```diff
@@ -585,34 +586,31 @@ class Parser : public gc {
...
         if (this->config.tokens && next.type !== Token.EOF) {
             this->tokens.push(this->convertToken(next));
         }
         */
-
-        return token;
     }
 
-    PassRefPtr<Scanner::ScannerResult> nextRegexToken()
+    void nextRegexToken(Scanner::ScannerResult* token)
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT token

File: src/parser/esprima_cpp/esprima.cpp [hunk 16]
```diff
@@ -634,7 +632,7 @@ class Parser : public gc {
         return n;
     }
 
-    MetaNode startNode(const RefPtr<Scanner::ScannerResult>& token)
+    MetaNode startNode(Scanner::ScannerResult* token)
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT token

File: src/parser/esprima_cpp/esprima.cpp [hunk 17]
```diff
@@ -1193,37 +1210,37 @@ class Parser : public gc {
...
+        Scanner::ScannerResult stricted;
         const char* message;
-        RefPtr<Scanner::ScannerResult> firstRestricted;
+        Scanner::ScannerResult firstRestricted;
         ParseParameterOptions()
         {
-            firstRestricted = nullptr;
-            stricted = nullptr;
+            //firstRestricted.reset();
+            //stricted.reset();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 18]
```diff
@@ -3109,66 +3154,67 @@ class Parser : public gc {
...
+            /*
             left->ref();
             stack.push_back(left.get());
-            token->ref();
-            stack.push_back(token.get());
+            //token->ref();
+            stack.push_back(token);
             right->ref();
             stack.push_back(right.get());
+            */
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 19]
```diff
@@ -3210,21 +3257,21 @@ class Parser : public gc {
                     right = *((ScanExpressionResult*)stack.back());
                     delete (ScanExpressionResult*)stack.back();
                     stack.pop_back();
-                    RefPtr<Scanner::ScannerResult> operator_ = (Scanner::ScannerResult*)stack.back();
-                    operator_->deref();
+                    Scanner::ScannerResult* operator_ = (Scanner::ScannerResult*)stack.back();
+                    //operator_->deref();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 20]
```diff
@@ -3210,21 +3257,21 @@ class Parser : public gc {
...
                     stack.push_back(new ScanExpressionResult(e));
                 }
 
                 // Shift.
-                token = this->nextToken();
+                this->nextToken(token);
                 token->prec = prec;
-                token->ref();
-                stack.push_back(token.get());
+                //token->ref();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 21]
```diff
@@ -3236,7 +3283,7 @@ class Parser : public gc {
             while (i > 1) {
                 expr = scanBinaryExpression(*((ScanExpressionResult*)stack[i - 2]), expr, (Scanner::ScannerResult*)stack[i - 1]);
                 delete (ScanExpressionResult*)stack[i - 2];
-                ((Scanner::ScannerResult*)stack[i - 1])->deref();
+                //((Scanner::ScannerResult*)stack[i - 1])->deref();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 22]
```diff
@@ -4433,7 +4486,9 @@ class Parser : public gc {
                 }
             } else if (this->matchKeyword(ConstKeyword) || this->matchKeyword(LetKeyword)) {
                 // TODO
-                this->throwUnexpectedToken(this->nextToken());
+                Scanner::ScannerResult* token = STACK_ALLOCA(sizeof(Scanner::ScannerResult), Scanner::ScannerResult);
+                this->nextToken(token);
+                this->throwUnexpectedToken(token);
                 /*
                 init = this->createNode();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: remove if not active codes

File: src/parser/Lexer.cpp [hunk 23]
```diff
@@ -631,30 +620,8 @@ Scanner::Scanner(::Escargot::Context* escargotContext, StringView code, ErrorHan
     , index(0)
     , lineNumber(((length > 0) ? 1 : 0) + startLine)
     , lineStart(startColumn)
-    , isPoolEnabled(true)
 {
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT escargotContext
ASSERT handler

File: src/parser/Lexer.cpp [hunk 24]
```diff
@@ -1300,14 +1269,13 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanNumericLiteral()
...
+    token->setResult(this, Token::NumericLiteralToken, ll, this->lineNumber, this->lineStart, start, this->index);
     if (startChar == '0' && length >= 2 && ll >= 1) {
-        ret->startWithZero = true;
+        token->startWithZero = true;
     }
-    return ret;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanStringLiteral()
+void Scanner::scanStringLiteral(Scanner::ScannerResult* token)
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT token

File: src/parser/Lexer.cpp [hunk 25]
```diff
@@ -857,9 +824,10 @@ uint16_t Scanner::octalToDecimal(char16_t ch, bool octal)
     return octal ? code : NON_OCTAL_VALUE;
 };
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanPunctuator(char16_t ch)
+void Scanner::scanPunctuator(Scanner::ScannerResult* token, char16_t ch)
 {
-    PassRefPtr<Scanner::ScannerResult> token = adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::PunctuatorToken, this->lineNumber, this->lineStart, this->index, this->index));
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 26]
```diff
@@ -1083,11 +1051,11 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanPunctuator(char16_t ch)
     }
 
     token->valuePunctuatorKind = kind;
-    return token;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanHexLiteral(size_t start)
+void Scanner::scanHexLiteral(Scanner::ScannerResult* token, size_t start)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 27]
```diff
@@ -1123,15 +1091,16 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanHexLiteral(size_t start)
...
         ASSERT(numberDouble == 0.0);
-        return adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index));
+        token->setResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index);
     }
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanBinaryLiteral(size_t start)
+void Scanner::scanBinaryLiteral(Scanner::ScannerResult* token, size_t start)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 28]
```diff
@@ -1158,11 +1127,12 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanBinaryLiteral(size_t start)
...
     }
 
-    return adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index));
+    token->setResult(this, Token::NumericLiteralToken, number, this->lineNumber, this->lineStart, start, this->index);
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanOctalLiteral(char16_t prefix, size_t start)
+void Scanner::scanOctalLiteral(Scanner::ScannerResult* token, char16_t prefix, size_t start)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 29]
```diff
@@ -1209,8 +1177,9 @@ bool Scanner::isImplicitOctalLiteral()
     return true;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanNumericLiteral()
+void Scanner::scanNumericLiteral(Scanner::ScannerResult* token)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 30]
```diff
@@ -1300,14 +1269,13 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanNumericLiteral()
...
     if (startChar == '0' && length >= 2 && ll >= 1) {
-        ret->startWithZero = true;
+        token->startWithZero = true;
     }
-    return ret;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanStringLiteral()
+void Scanner::scanStringLiteral(Scanner::ScannerResult* token)
 {
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 31]
```diff
@@ -1375,17 +1343,15 @@ PassRefPtr<Scanner::ScannerResult> Scanner::scanStringLiteral()
     }
 
     if (isPlainCase) {
-        StringView str(this->source, start + 1, this->index - 1);
-        auto ret = adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, Token::StringLiteralToken, str, this->lineNumber, this->lineStart, start, this->index, true));
-        ret->octal = octal;
-        ret->plain = true;
-        return ret;
+        StringView* str = new StringView(this->source, start + 1, this->index - 1);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(str != nullptr);

File: src/parser/Lexer.cpp [hunk 32]
```diff
@@ -473,42 +473,31 @@ void ErrorHandler::throwError(size_t index, size_t line, size_t col, String* des
...
+    return sv->string();
 }
 
-StringView Scanner::ScannerResult::valueStringLiteral()
+StringView* Scanner::ScannerResult::valueStringLiteral()
 {
     if (this->type == Token::KeywordToken && !this->hasKeywordButUseString) {
         AtomicString as = keywordToString(this->scanner->escargotContext, this->valueKeywordKind);
-        return StringView(as.string(), 0, as.string()->length());
+        return new StringView(as.string(), 0, as.string()->length());
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: before return, check nullity of ``new`` with ASSERT

File: src/parser/Lexer.cpp [hunk 33]
```diff
@@ -620,7 +609,7 @@ void Scanner::ScannerResult::constructStringLiteral()
     } else {
         newStr = new UTF16String(stringUTF16.data(), stringUTF16.length());
     }
-    this->valueStringLiteralData = StringView(newStr, 0, newStr->length());
+    this->valueStringLiteralData = new StringView(newStr, 0, newStr->length());
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: before return, check nullity of ``new`` with ASSERT

File: src/parser/Lexer.cpp [hunk 34]
```diff
@@ -762,10 +729,10 @@ StringView Scanner::getIdentifier()
         }
     }
 
-    return StringView(this->source, start, this->index);
+    return new StringView(this->source, start, this->index);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: before return, check nullity of ``new`` with ASSERT

File: src/parser/Lexer.cpp [hunk 35]
```diff
@@ -829,7 +796,7 @@ StringView Scanner::getComplexIdentifier()
     }
 
     String* str = new UTF16String(id.data(), id.length());
-    return StringView(str, 0, str->length());
+    return new StringView(str, 0, str->length());
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: before return, check nullity of ``new`` with ASSERT

File: src/parser/Lexer.cpp [hunk 36]
```diff
@@ -1403,8 +1369,9 @@ bool Scanner::isFutureReservedWord(const StringView& id)
     return false;
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanTemplate(bool head)
+void Scanner::scanTemplate(Scanner::ScannerResult* token, bool head)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 37]
```diff
@@ -1616,8 +1583,9 @@ String* Scanner::scanRegExpFlags()
     return new UTF16String(flags.data(), flags.length());
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::scanRegExp()
+void Scanner::scanRegExp(Scanner::ScannerResult* token)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 38]
```diff
@@ -1802,25 +1769,26 @@ static ALWAYS_INLINE KeywordKind isKeyword(const StringBufferAccessData& data)
     return NotKeyword;
 }
 
-ALWAYS_INLINE PassRefPtr<Scanner::ScannerResult> Scanner::scanIdentifier(char16_t ch0)
+ALWAYS_INLINE void Scanner::scanIdentifier(Scanner::ScannerResult* token, char16_t ch0)
 {
+    ASSERT(token);
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.cpp [hunk 39]
```diff
@@ -1835,13 +1803,14 @@ ALWAYS_INLINE PassRefPtr<Scanner::ScannerResult> Scanner::scanIdentifier(char16_
         type = Token::IdentifierToken;
     }
 
-    return adoptRef(new (createScannerResult()) Scanner::ScannerResult(this, type, id, this->lineNumber, this->lineStart, start, this->index, id.string() == this->source.string()));
+    token->setResult(this, type, id, this->lineNumber, this->lineStart, start, this->index, id->string() == this->source.string());
 }
 
-PassRefPtr<Scanner::ScannerResult> Scanner::lex()
+void Scanner::lex(Scanner::ScannerResult* token)
 {
```
Comment (yichoi) [inline src/parser/Lexer.cpp]: ASSERT(token != nullptr);

File: src/parser/Lexer.h [hunk 40]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
-            , plain(false)
-            , hasKeywordButUseString(false)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueNumber(0)
+        void setResult(Scanner* scanner, Token type, size_t lineNumber, size_t lineStart, size_t start, size_t end)
         {
+            ASSERT(scanner);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(scanner != nullptr);

File: src/parser/Lexer.h [hunk 41]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
-            , plain(plain)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueStringLiteralData(valueString)
+        void setResult(Scanner* scanner, Token type, StringView* valueString, size_t lineNumber, size_t lineStart, size_t start, size_t end, bool plain)
         {
+            ASSERT(scanner);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(token != nullptr);

File: src/parser/Lexer.h [hunk 42]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
+            ASSERT(valueString);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(valueString != nullptr);

File: src/parser/Lexer.h [hunk 43]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
-            , plain(false)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueNumber(value)
+        void setResult(Scanner* scanner, Token type, double value, size_t lineNumber, size_t lineStart, size_t start, size_t end)
         {
+            ASSERT(scanner);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(scanner != nullptr);

File: src/parser/Lexer.h [hunk 44]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
-            , plain(false)
-            , hasKeywordButUseString(true)
-            , lineNumber(lineNumber)
-            , lineStart(lineStart)
-            , start(start)
-            , end(end)
-            , valueTemplate(value)
+        void setResult(Scanner* scanner, Token type, ScanTemplteResult* value, size_t lineNumber, size_t lineStart, size_t start, size_t end)
         {
+            ASSERT(scanner);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(scanner != nullptr);

File: src/parser/Lexer.h [hunk 45]
```diff
@@ -233,95 +255,104 @@ class Scanner : public gc {
...
+            ASSERT(value);
```
Comment (yichoi) [inline src/parser/Lexer.h]: ASSERT(value != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 46]
```diff
@@ -517,7 +516,7 @@ class Parser : public gc {
     }
 
     // Throw an exception because of an unexpected token.
-    void throwUnexpectedToken(RefPtr<Scanner::ScannerResult> token, const char* message = nullptr)
+    void throwUnexpectedToken(Scanner::ScannerResult* token, const char* message = nullptr)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 47]
```diff
@@ -585,34 +587,31 @@ class Parser : public gc {
...
             this->tokens.push(this->convertToken(next));
         }
         */
-
-        return token;
     }
 
-    PassRefPtr<Scanner::ScannerResult> nextRegexToken()
+    void nextRegexToken(Scanner::ScannerResult* token)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 48]
```diff
@@ -634,7 +633,7 @@ class Parser : public gc {
         return n;
     }
 
-    MetaNode startNode(const RefPtr<Scanner::ScannerResult>& token)
+    MetaNode startNode(Scanner::ScannerResult* token)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 49]
```diff
@@ -942,28 +943,28 @@ class Parser : public gc {
...
             }
             this->lastMarker.index = this->startMarker.index;
             this->lastMarker.lineNumber = this->startMarker.lineNumber;
             this->lastMarker.lineStart = this->startMarker.lineStart;
         }
     }
 
-    IdentifierNode* finishIdentifier(PassRefPtr<Scanner::ScannerResult> token, bool isScopeVariableName)
+    IdentifierNode* finishIdentifier(Scanner::ScannerResult* token, bool isScopeVariableName)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 50]
```diff
@@ -973,18 +974,18 @@ class Parser : public gc {
         return ret;
     }
 
-    ScanExpressionResult finishScanIdentifier(PassRefPtr<Scanner::ScannerResult> token, bool isScopeVariableName)
+    ScanExpressionResult finishScanIdentifier(Scanner::ScannerResult* token, bool isScopeVariableName)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 51]
```diff
@@ -1193,37 +1211,35 @@ class Parser : public gc {
...
         {
-            firstRestricted = nullptr;
-            stricted = nullptr;
-            message = nullptr;
         }
     };
 
-    void validateParam(ParseParameterOptions& options, const RefPtr<Scanner::ScannerResult>& param, AtomicString name)
+    void validateParam(ParseParameterOptions& options, const Scanner::ScannerResult* param, AtomicString name)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(param != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 52]
```diff
@@ -2421,15 +2457,16 @@ class Parser : public gc {
         }
     }
 
-    bool isIdentifierName(RefPtr<Scanner::ScannerResult> token)
+    bool isIdentifierName(Scanner::ScannerResult* token)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 53]
```diff
@@ -3030,7 +3072,7 @@ class Parser : public gc {
     // ECMA-262 12.12 Binary Bitwise Operators
     // ECMA-262 12.13 Binary Logical Operators
 
-    int binaryPrecedence(const RefPtr<Scanner::ScannerResult>& token)
+    int binaryPrecedence(const Scanner::ScannerResult* token)
     {
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: ASSERT(token != nullptr);

File: src/parser/esprima_cpp/esprima.cpp [hunk 54]
```diff
@@ -3195,12 +3233,12 @@ class Parser : public gc {
 
             std::vector<void*, GCUtil::gc_malloc_ignore_off_page_allocator<void*>> stack;
             stack.push_back(new ScanExpressionResult(left));
-            token->ref();
-            stack.push_back(token.get());
+            //token->ref();
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: Remove if not active codes

File: src/parser/esprima_cpp/esprima.cpp [hunk 55]
```diff
@@ -5562,20 +5608,21 @@ class Parser : public gc {
         this->consumeSemicolon();
 
         if (isLiteral) {
-            return this->finalize(node, new DirectiveNode(asExpressionNode(expr).get(), directiveValue));
+            return this->finalize(node, new DirectiveNode(asExpressionNode(expr).get(), *directiveValue));
```
Comment (yichoi) [inline src/parser/esprima_cpp/esprima.cpp]: before return, check nullity of ``new`` with ASSERT

File: src/parser/Lexer.cpp (hunk 31 above)
Comment (clover2123) [reply src/parser/Lexer.cpp]: `StringView` is allocated by GC. If memory alloc is failed, GC will throw or warn error.

File: src/parser/Lexer.cpp (hunk 32 above)
Comment (clover2123) [reply src/parser/Lexer.cpp]: Same as above.

File: src/parser/Lexer.cpp (hunk 33 above)
Comment (clover2123) [reply src/parser/Lexer.cpp]: Same as above.

File: src/parser/esprima_cpp/esprima.cpp (hunk 55 above)
Comment (clover2123) [reply src/parser/esprima_cpp/esprima.cpp]: ASTNode is also allocated by GC. So, if there is any problem in allocation, GC will notify us.

File: src/parser/esprima_cpp/esprima.cpp (hunk 22 above)
Comment (clover2123) [reply src/parser/esprima_cpp/esprima.cpp]: Our parser is based on the esprima parser. This comment is related with esprima code which is left to implement es6 features later.

**Comment by clover2123** [comment]
@yichoi I've updated the patch as you commented except some new allocations and esprima code comments (I replied these comments above). Thanks for the review.

**Pull request merged**